### PR TITLE
Statistics: keep rows ordered during sync insert

### PR DESCRIPTION
**Pull request by snelg**
Honestly, I might be literally the only person who cares about this.

During the Statistics sync, if new rows are inserted then they're not added to the database in any particular order. That creates absolutely no problems for anything within KOReader itself.
But if you have more than one reader and you sync each of them at semi-regular intervals, then if you do a sqlite `.dump` of the databases and `diff` those text files, the lines are in scrambled order from each other.

So this PR has zero effect on any visible behavior within KOReader. It only helps weirdos like me doing obscure text processing on dumps of the statistics database :)

- - -
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/10952)

### CODE REVIEW

**Review by Frenzie — APPROVED** [review]
Hopefully it doesn't add too much extra processing time though? :-)

**Comment by snelg** [comment]
> Hopefully it doesn't add too much extra processing time though? :-)

No, doesn't add much extra processing time. That `start_time` field is indexed.

**Comment by snelg** [comment]
Huh, that's unexpected. I just ran some tests, and adding that `ORDER BY` clause slightly *speeds up* the upsert. Not enough to be human-noticeable in most practical cases, but the difference is quantifiable.

Ah, I think I understand why, at least vaguely:
 - Without the ORDER clause, `EXPLAIN QUERY PLAN` shows that it scans through the book mapping temp table, and then searches the page_stat_data table for each row from the mapping table
 - With the added ORDER clause, it does it the other way around, i.e. scans the page_stat_data table and searches the mapping table for each row from page_stat_data

Since the page_stat_data is much larger than the book mapping temp table, it's more efficient to do that second variation
 
 In short, yet again I am getting way too interested in a tiny detail that no one else is going to care about.

**Comment by snelg** [comment]
Sigh. Never mind; I ran those performance tests on a virtual device. But when I ran them on a physical device (a Kobo Forma in this case), it horribly affected performance.

So I'm canceling this PR and I'll deal with my little offline processing stuff on my own.

**Comment by Frenzie** [comment]
I guess it's I/O bound? The CPU's a lot slower but I wouldn't generally expect huge differences in the performance characteristics for something like this (though you never know without testing).

**Comment by snelg** [comment]
Double-sigh. The performance issue was probably actually a network issue. I was testing the whole "synchronize stats" process, not just that one sql statement. That's when the performance dropped. But everything worked fine when I tried again this morning. *shrug*
Either way, it made me nervous about potential unwanted consequences. And as I said from the start, probably literally no one else on earth actually cares about this, so I'll handle it on my own.

**Comment by Frenzie** [comment]
> a network issue

The slowest I/O! :laughing: 

> probably literally no one else on earth actually cares about this, so I'll handle it on my own.

Like I said it seems perfectly fine to me to add the change. It's just adding a little sort by, not adding an entirely new seemingly unused module or whatever. ;-)

**Comment by poire-z** [comment]
> requested a review from [poire-z](https://github.com/poire-z)

I think this was more about @weijiuqiao 's sync work than what I'm confortable with :)

As for the regular stats insertion, we don't order anything explicitely in the plugin, but I assume that from our Lua table, they will get inserted in some proper order (that will make our other SQL reading queries get their stuff without sqlite having to do much additional reordering?):
https://github.com/koreader/koreader/blob/4875f63f09a98e69575e0d06ebfc8561d1dff438/plugins/statistics.koplugin/main.lua#L960-L980

**Comment by snelg** [comment]
> As for the regular stats insertion, we don't order anything explicitely in the plugin, but I assume that from our Lua table, they will get inserted in some proper order (that will make our other SQL reading queries get their stuff without sqlite having to do much additional reordering?):

Yea, sorry, I tried to make this as clear as possible, but it's such a tiny technical detail that it's hard to explain without getting deep into sqlite's inner workings.

But the most important point is that the existing code works perfectly fine for everything KOReader does. There is absolutely no problem inside KOReader I was trying to "fix."

It's just an obscure detail about the output if you do a sqlite ".dump" command to generate a SQL text file. Which most normal people would never have reason to do. Or even most less-normal people ;)

**Comment by snelg** [comment]
Oh, and more importantly I now have a workaround from my end of things, so I definitely don't need this PR re-opened or anything. I've already wasted enough of everyone else's time.

**Comment by Frenzie** [comment]
> It's just an obscure detail about the output if you do a sqlite ".dump" command to generate a SQL text file. Which most normal people would never have reason to do. Or even most less-normal people ;)

It's true that I'd use either sqlite3 or DB Browser for SQLite, though it still feels like something I might do. :-P
